File: apps/desktop/src/shared/logger/redactor.test.ts
// The canonical Redactor tests live alongside the package at
// `apps/desktop/src/shared/redact/redact.test.ts` (per hp-je1p). The tests
// below cover the logger-side integration: redactLogEntry handles the
// LogEntry envelope correctly and back-compat re-exports work.

import { describe, expect, test } from "bun:test";
import { Redactor, redactLogEntry } from "./redactor.ts";
import type { LogEntry } from "./types.ts";

describe("logger.Redactor (re-export from shared/redact)", () => {
  test("re-export still scrubs known patterns via redactText", () => {
    const r = new Redactor();
    const { redacted, events } = r.redactText("logger", "msg", "token=sk-abcdef0123456789ABCDEF0123456789");
    expect(redacted).not.toContain("sk-abcdef");
    expect(events.length).toBeGreaterThan(0);
  });
});

describe("redactLogEntry", () => {
  test("scrubs msg + fields, leaves envelope columns", () => {
    const r = new Redactor();
    const input: LogEntry = {
      ts: "2026-05-04T00:00:00Z",
      level: "info",
      msg: "calling with sk-abcdef0123456789ABCDEF0123456789",
      component: "desktop.main",
      jobId: "job-1",
      fields: {
        argv: ["--token", "ghp_abcdefghijklmnopqrstuvwxyz0123456789"],
        benign: "just a value",
      },
    };
    const { entry, events } = redactLogEntry(r, input);
    expect(entry.msg).not.toContain("sk-abcdef");
    expect(entry.component).toBe("desktop.main");
    expect(entry.jobId).toBe("job-1");
    const argv = entry.fields?.argv as string[];
    expect(argv[1]).not.toContain("ghp_");
    expect(entry.fields?.benign).toBe("just a value");
    expect(events.length).toBeGreaterThanOrEqual(2);
    // All events stamped as logger surface.
    for (const e of events) {
      expect(e.redactor).toBe("logger");
    }
  });

  test("empty entry passes through unchanged", () => {
    const r = new Redactor();
    const input: LogEntry = {
      ts: "x",
      level: "info",
      msg: "",
      component: "test",
    };
    const { entry, events } = redactLogEntry(r, input);
    expect(entry).toEqual(input);
    expect(events).toEqual([]);
  });

  test("does not mutate input entry", () => {
    const r = new Redactor();
    const input: LogEntry = {
      ts: "x",
      level: "info",
      msg: "leak sk-abcdef0123456789ABCDEF0123456789",
      component: "test",
    };
    const before = { ...input };
    redactLogEntry(r, input);
    expect(input.msg).toBe(before.msg);
  });

  test("field context paths propagate", () => {
    const r = new Redactor();
    const input: LogEntry = {
      ts: "x",
      level: "info",
      msg: "ok",
      component: "test",
      fields: {
        token: "sk-abcdef0123456789ABCDEF0123456789",
      },
    };
    const { events } = redactLogEntry(r, input);
    expect(events[0].context).toBe("fields.token");
  });
});

File: apps/daemon/internal/logger/redactor_test.go
package logger

import (
	"strings"
	"testing"

	"github.com/hoopoe-cockpit/hoopoe/apps/daemon/internal/redaction"
)

// The canonical redactor tests live in `apps/daemon/internal/redaction/`
// (per hp-je1p). The logger-side tests below cover the Entry integration
// (RedactEntry) and back-compat re-exports.

func TestNewRedactorRetainsCanonicalPatterns(t *testing.T) {
	r := NewRedactor()
	if r == nil {
		t.Fatal("NewRedactor returned nil")
	}
	out, events := r.RedactText(redaction.SurfaceLogger, "test", "token=sk-abcdef0123456789ABCDEF0123456789")
	if strings.Contains(out, "sk-abcdef") {
		t.Errorf("redactor did not scrub: %s", out)
	}
	if len(events) == 0 {
		t.Errorf("expected events, got none")
	}
}

// RedactEntry mutates msg + every value in fields. Envelope columns
// (component, jobId, ...) are NOT redacted.
func TestRedactEntryScrubsMsgAndFields(t *testing.T) {
	r := NewRedactor()
	e := &Entry{
		Msg: "calling backend with sk-abcdef0123456789ABCDEF0123456789",
		Fields: map[string]any{
			"argv":   []any{"--token", "ghp_abcdefghijklmnopqrstuvwxyz0123456789"},
			"benign": "just a value",
		},
	}
	events := RedactEntry(r, e)
	if strings.Contains(e.Msg, "sk-abcdef") {
		t.Errorf("msg not redacted: %s", e.Msg)
	}
	argv, ok := e.Fields["argv"].([]any)
	if !ok {
		t.Fatalf("argv not array")
	}
	if v, _ := argv[1].(string); strings.Contains(v, "ghp_") {
		t.Errorf("argv token not redacted: %v", v)
	}
	if e.Fields["benign"] != "just a value" {
		t.Errorf("benign field mutated: %v", e.Fields["benign"])
	}
	if len(events) < 2 {
		t.Errorf("expected ≥2 events, got %d", len(events))
	}
}

func TestRedactEntryNilSafe(t *testing.T) {
	if events := RedactEntry(nil, &Entry{Msg: "x"}); events != nil {
		t.Errorf("nil redactor should yield nil events")
	}
	r := NewRedactor()
	if events := RedactEntry(r, nil); events != nil {
		t.Errorf("nil entry should yield nil events")
	}
}

func TestRedactEntryEmptyEntryIsNoOp(t *testing.T) {
	r := NewRedactor()
	e := &Entry{Msg: "", Fields: nil}
	events := RedactEntry(r, e)
	if len(events) != 0 {
		t.Errorf("empty entry produced events: %v", events)
	}
}

// Surface label flows through into the trace events so audit-log replay
// tools can distinguish where the redaction happened.
func TestRedactEntryStampsLoggerSurface(t *testing.T) {
	r := NewRedactor()
	e := &Entry{
		Msg:    "leak sk-abcdef0123456789ABCDEF0123456789",
		Fields: nil,
	}
	events := RedactEntry(r, e)
	if len(events) == 0 {
		t.Fatal("expected events")
	}
	if events[0].Redactor != string(redaction.SurfaceLogger) {
		t.Errorf("expected redactor=%s, got %s", redaction.SurfaceLogger, events[0].Redactor)
	}
}
